internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,371 @@
+	define parent: [document, module]
+	define viewer: [user] or owner or viewer from parent
+type folder
+relations
+	define owner: [user] or owner from parent
+	define parent: [module, folder]
+	define viewer: [user] or owner or viewer from parent
+type document
+relations
+	define owner: [user] or owner from parent
+	define parent: [folder, document]
+	define viewer: [user] or owner or viewer from parent`)
+
+	err = ds.WriteAuthorizationModel(context.Background(), storeID, model)
+	require.NoError(t, err)
+
+	ctx := typesystem.ContextWithTypesystem(
+		context.Background(),
+		typesystem.New(model),
+	)
+	ctx = storage.ContextWithRelationshipTupleReader(ctx, ds)
+
+	singleflightCheckResolver := NewSingleflightCheckResolver()
+	localCheckResolver := NewLocalChecker()
+
+	singleflightCheckResolver.SetDelegate(localCheckResolver)
+	localCheckResolver.SetDelegate(singleflightCheckResolver)
+
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+		singleflightCheckResolver.Close()
+		localCheckResolver.Close()
+	})
+
+	tupleKeys := []*openfgav1.TupleKey{
+		tuple.NewTupleKey("module:a", "viewer", "user:anne"),
+		tuple.NewTupleKey("module:b", "viewer", "user:anne"),
+		tuple.NewTupleKey("folder:a", "viewer", "user:anne"),
+		tuple.NewTupleKey("folder:b", "viewer", "user:anne"),
+		tuple.NewTupleKey("document:a", "viewer", "user:anne"),
+		tuple.NewTupleKey("document:b", "viewer", "user:anne"),
+	}
+
+	for _, tupleKey := range tupleKeys {
+		resp, err := singleflightCheckResolver.ResolveCheck(ctx, &ResolveCheckRequest{
+			StoreID:              storeID,
+			AuthorizationModelID: model.GetId(),
+			TupleKey:             tupleKey,
+			ResolutionMetadata: &ResolutionMetadata{
+				Depth: 25,
+			},
+		})
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	}
+}
+
+// TestSingleflightCheckResolver_ContextCancellation is a test to assert
+// that if two FGA queries that have overlapping nested subproblems share
+// different contexts, and one is cancelled out-of-band from the other, then
+// they should both resolve with non-nil errors. If they do not, then that
+// indicates incorrect handling of context cancellation in the call chain.
+func TestSingleflightCheckResolver_ContextCancellation(t *testing.T) {
+	storeID := ulid.Make().String()
+
+	ds := memory.New()
+	err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+		tuple.NewTupleKey("folder:2", "folder_reader", "user:anne"),
+		tuple.NewTupleKey("folder:2", "allowed", "folder:1"),
+		tuple.NewTupleKey("folder:4", "parent", "folder:2"),
+		tuple.NewTupleKey("folder:4", "allowed", "user:anne"),
+	})
+	require.NoError(t, err)
+
+	model := testutils.MustTransformDSLToProtoWithID(`model
+  schema 1.1
+
+type user
+type folder
+relations
+	define parent: [folder]
+	define folder_reader: [user] or folder_reader from parent
+	define blocked: [user]
+	define allowed: [user] or allowed from parent
+	define reader: folder_reader and allowed
+ `)
```

Comment:
can we simplify this model? even if you have a model with simply `define reader: [user]` and you spin up two goroutines doing the same `Check(folder:1, reader, user:x)`, one of them should _mostly_ wait on the other one to finish, right?